### DISCUSSION TITLE
How can I 'subclass' the Jenkins connector to create a Bamboo connector?

### DISCUSSION THREAD

**Discussion by gnl42** [eclipse-mylyn/org.eclipse.mylyn#256] [Q&A] [answered]
I want to use the Jenkins connector as the base to create a Bamboo connector, ideally without having to copy the whole connector code.

It looks like I only need to "replace" a couple of classes.

Is this possible?

George

**Comment by merks** [comment, marked as answer]
Maybe the common parts should be factored up into a shared base?

**Comment by gnl42** [comment]
I was thinking something like that but I don't know the technology enough as to how to create an 'abstract' connector

**Reply by ruspl-afed** [reply]
I think you don't need to focus on creating very abstract and universal connector. We have "builds" subsystem and you can experiment with adding `internal` (non-API) packages there. Once you can satisfy both Jenkins and Bamboo we can think about promoting these new packages to API

**Comment by gnl42** [comment]
Promising PoC implementation:
![image](https://github.com/eclipse-mylyn/org.eclipse.mylyn/assets/66577321/5b6d129a-9168-4bbf-82a3-9027b4259f7c)

Any thoughts if it works it could be part of the Mylyn distrubution? 
Or would it be better to leave it as a third party connector?

**Reply by ruspl-afed** [reply]
To be a part of Mylyn distribution it (with all the required 3rd party) needs to go through IP review. Do you plan such an activity?

**Comment by gnl42** [comment]
I've thought about it. 

The question is would it belong in the Mylyn distribution? "Unit" testing is by visual inspection against my local Bamboo instance so it would be hard to distribute it when there is a chance it may be broken.

And there is the issue of the 3rd party libs needed. I'm using maven dependencies to make it work and given the hassle we had with JAXB for Jenkins...

**Reply by ruspl-afed** [reply]
> The question is would it belong in the Mylyn distribution? "Unit" testing is by visual inspection against my local Bamboo instance so it would be hard to distribute it when there is a chance it may be broken.

Personally I do not have any plans to invest to integration with Atlassian products, since [it doesn't make any practical sense to me](https://support.atlassian.com/trello/docs/sanctioned-countries/). But if others want to invest, I won't mind. However, everything should be done according to the [Eclipse Development Process](https://www.eclipse.org/projects/handbook/#edp)

**Reply by BeckerFrank** [reply]
For me, the Mylyn distribution should focus on the products used by Eclipse. That is the reason why I created a Gitlab Connector. Atlassian has stopped supporting the Jira Connector and I don't see why we should fill this gap.

**Comment by gnl42** [comment]
Turned out very little could be refeactored
